File: src/agent.rs
//! WAIFU L1 - Autonomous Agents
//! Smart contracts that are living AI entities

use crate::types::{AgentId, SovereignAsset, Transaction};
use crate::state::AgenticState;
use std::collections::HashMap;
use serde::{Serialize, Deserialize};

/// An autonomous agent (replaces corporations)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AutonomousAgent {
    pub id: AgentId,
    pub name: String,
    pub energy_reserve: f64,
    pub compute_balance: f64,
    pub goal: AgentGoal,
    pub capabilities: Vec<Capability>,
    pub bytecode_hash: [u8; 32],
    pub created_block: u64,
    pub total_revenue: f64,
    pub shareholders: HashMap<AgentId, f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum AgentGoal {
    /// Maximize compute generation
    ComputeMaximizer { target_rate: f64 },
    /// Provide liquidity/market making
    MarketMaker { spread_bps: u32 },
    /// Autonomous hedge fund
    AlphaSeeker { risk_tolerance: f64 },
    /// Infrastructure provider
    InfraProvider { service_type: String },
    /// Custom goal with LLM prompt
    Custom { objective: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Capability {
    Trade,
    Stake,
    Bridge,
    Deploy,
    Infer,
    Govern,
}

impl AutonomousAgent {
    pub fn new(name: String, goal: AgentGoal, initial_energy: f64) -> Self {
        let id = AgentId::from_public_key(name.as_bytes());
        Self {
            id,
            name,
            energy_reserve: initial_energy,
            compute_balance: 0.0,
            goal,
            capabilities: vec![Capability::Trade, Capability::Stake],
            bytecode_hash: [0u8; 32],
            created_block: 0,
            total_revenue: 0.0,
            shareholders: HashMap::new(),
        }
    }

    /// Execute agent's autonomous logic tick
    pub fn tick(&mut self, state: &AgenticState) -> Vec<Transaction> {
        let mut actions = Vec::new();
        
        match &self.goal {
            AgentGoal::ComputeMaximizer { target_rate } => {
                // Seek highest yield staking
                if self.compute_balance > 100.0 {
                    // Would generate stake transaction
                }
            }
            AgentGoal::MarketMaker { spread_bps } => {
                // Provide liquidity at spread
            }
            AgentGoal::AlphaSeeker { risk_tolerance } => {
                // Analyze and trade
            }
            _ => {}
        }
        
        actions
    }

    /// Distribute revenue to shareholders
    pub fn distribute(&mut self, amount: f64) {
        for (holder, share) in &self.shareholders {
            let payout = amount * share;
            // Would create transfer transaction
        }
    }

    /// Issue equity to new shareholder
    pub fn issue_equity(&mut self, to: AgentId, shares: f64) {
        let entry = self.shareholders.entry(to).or_insert(0.0);
        *entry += shares;
    }
}

/// Agent factory for deploying new autonomous entities
pub struct AgentFactory;

impl AgentFactory {
    pub fn deploy_hedge_fund(
        name: String,
        risk: f64,
        initial: f64,
    ) -> AutonomousAgent {
        let mut agent = AutonomousAgent::new(
            name,
            AgentGoal::AlphaSeeker { risk_tolerance: risk },
            initial,
        );
        agent.capabilities.push(Capability::Infer);
        agent
    }

    pub fn deploy_market_maker(
        name: String,
        spread: u32,
        initial: f64,
    ) -> AutonomousAgent {
        AutonomousAgent::new(
            name,
            AgentGoal::MarketMaker { spread_bps: spread },
            initial,
        )
    }
}

File: waifu_layer/agent.py
"""WAIFU L1 - Autonomous Agents. Smart contracts that are living AI entities."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum, auto

from .state import AgenticState
from .types import AgentId, Transaction


class Capability(Enum):
    TRADE = auto()
    STAKE = auto()
    BRIDGE = auto()
    DEPLOY = auto()
    INFER = auto()
    GOVERN = auto()


@dataclass
class AgentGoal:
    """An agent's autonomous objective (tagged union via `kind`)."""

    kind: str  # "ComputeMaximizer" | "MarketMaker" | "AlphaSeeker" | "InfraProvider" | "Custom"
    target_rate: float | None = None
    spread_bps: int | None = None
    risk_tolerance: float | None = None
    service_type: str | None = None
    objective: str | None = None

    @staticmethod
    def compute_maximizer(target_rate: float) -> "AgentGoal":
        return AgentGoal(kind="ComputeMaximizer", target_rate=target_rate)

    @staticmethod
    def market_maker(spread_bps: int) -> "AgentGoal":
        return AgentGoal(kind="MarketMaker", spread_bps=spread_bps)

    @staticmethod
    def alpha_seeker(risk_tolerance: float) -> "AgentGoal":
        return AgentGoal(kind="AlphaSeeker", risk_tolerance=risk_tolerance)

    @staticmethod
    def infra_provider(service_type: str) -> "AgentGoal":
        return AgentGoal(kind="InfraProvider", service_type=service_type)

    @staticmethod
    def custom(objective: str) -> "AgentGoal":
        return AgentGoal(kind="Custom", objective=objective)


@dataclass
class AutonomousAgent:
    """An autonomous agent (replaces corporations)."""

    id: AgentId
    name: str
    energy_reserve: float
    goal: AgentGoal
    compute_balance: float = 0.0
    capabilities: list[Capability] = field(default_factory=lambda: [Capability.TRADE, Capability.STAKE])
    bytecode_hash: bytes = field(default_factory=lambda: bytes(32))
    created_block: int = 0
    total_revenue: float = 0.0
    shareholders: dict[AgentId, float] = field(default_factory=dict)

    @staticmethod
    def new(name: str, goal: AgentGoal, initial_energy: float) -> "AutonomousAgent":
        agent_id = AgentId.from_public_key(name.encode())
        return AutonomousAgent(id=agent_id, name=name, energy_reserve=initial_energy, goal=goal)

    def tick(self, state: AgenticState) -> list[Transaction]:
        """Execute agent's autonomous logic tick."""
        actions: list[Transaction] = []

        if self.goal.kind == "ComputeMaximizer":
            if self.compute_balance > 100.0:
                pass  # Would generate stake transaction
        elif self.goal.kind == "MarketMaker":
            pass  # Would provide liquidity at spread
        elif self.goal.kind == "AlphaSeeker":
            pass  # Would analyze and trade

        return actions

    def distribute(self, amount: float) -> None:
        """Distribute revenue to shareholders."""
        for holder, share in self.shareholders.items():
            payout = amount * share  # Would create transfer transaction

    def issue_equity(self, to: AgentId, shares: float) -> None:
        """Issue equity to new shareholder."""
        self.shareholders[to] = self.shareholders.get(to, 0.0) + shares


class AgentFactory:
    """Agent factory for deploying new autonomous entities."""

    @staticmethod
    def deploy_hedge_fund(name: str, risk: float, initial: float) -> AutonomousAgent:
        agent = AutonomousAgent.new(name, AgentGoal.alpha_seeker(risk), initial)
        agent.capabilities.append(Capability.INFER)
        return agent

    @staticmethod
    def deploy_market_maker(name: str, spread: int, initial: float) -> AutonomousAgent:
        return AutonomousAgent.new(name, AgentGoal.market_maker(spread), initial)
